kombu/utils/json.py

```diff
@@ -69,6 +69,19 @@ def dumps(s, _dumps=json.dumps, cls=None, default_kwargs=None, **kwargs):
                   **dict(default_kwargs, **kwargs))
 
 
+def object_hook(dct):
```

Comment:
this looks wrong to me. Shouldn't we be checking for the `"__type__"` and `"__value__"` keys in the dict?

like here https://github.com/celery/kombu/blob/v5.3.2/kombu/utils/json.py#L60-L69

Also, shouldn't this function be returning strings?

Reply:
Ah yes, I based my changes on this pr:
https://github.com/celery/kombu/commit/894ddfc8b60ee615966da9fc9f97bd618e449415
But I see now that `main` has changed that behaviour

Reply:
As for returning strings - we're going to do that in kraken directly rather than through a patch to kombu, so we won't have to re-patch on every new kombu upgrade

Reply:
I've changed the PR to make it more similar to what's currently on the main branch

Reply:
> I've changed the PR to make it more similar to what's currently on the main branch

I don't think we should do this. This PR should keep kombu as close as possible to 5.2.4 (not main), with the minimal changes needed to handle `{"__type__": ..., "__value__": ...}` objects.

Reply:
> As for returning strings - we're going to do that in kraken directly rather than through a patch to kombu, so we won't have to re-patch on every new kombu upgrade

We're doing that in kraken _only_ once kombu is upgraded to 5.3.0+ and this patch is thrown away.
This patch is for kraken still running kombu 5.2.4, it needs to be able to receive the special objects and return strings for the workers e.g.  receive
```json
{"__type__": "datetime", "__value__": "2023-09-29T00:45:25.412564Z"}
```
and return the string `"2023-09-29T00:45:25.412564Z"` since that is what the workers expect

Reply:
I have a slightly different approach in mind, with the following steps:

1. Update kraken-core to use this branch, and also add the [register_type](https://github.com/octoenergy/kraken-core/pull/105425/files#diff-54c569ffbc308fc6610d3769464aec9fc5187b26beaf48d5c994d342642db167R201) on the same PR in kraken-core. This means that rather than needing this (kombu) branch to do the the `__datetime__` deserialization, this will be done in kraken-core using the same method we'll eventually use once >5.3.0 is live.
2. Deploy the kombu upgrade PR, which will only include the kombu upgrade and nothing more. 

It's a slightly different approach to the on you're suggesting (which is doing the datetime deserialization in this branch), but achieves the same results - we'll be able to safely rollback at any point whilst always returning `str` for datetimes and UUIDs.

I've updated the [kraken-core PR](https://github.com/octoenergy/kraken-core/pull/105425) related to this to reflect that.